torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -307,8 +321,19 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
 void ProcessGroupNCCL::WorkNCCL::synchronizeStreams() {
   for (size_t i = 0; i < devices_.size(); ++i) {
     auto currentStream = at::cuda::getCurrentCUDAStream(devices_[i].index());
+
     // Block the current stream on the NCCL stream
     cudaEvents_[i].block(currentStream);
+    // Enqueue guard function as a callback on the current stream to throw
+    // user-stream exception upon error.
+    if (!blockingWait_) {
+      cudaStreamAddCallback(currentStream, errorGuard, this, 0);
```

Comment:
IIRC @sinannasir mentioned that this callback leads to 30% perf regression due to H2D comm. Is that correct? cc @sinannasir 

Shall we run some benchmark to make sure this does not introduce the same regression here?

Reply:
@mrshenli We actually remove the `cudaStreamAddCallback` in the last diff in this stack and replace it with a `cudaEventQuery`-based solution (the stream callback was the original solution, and I simply added the change to the top of the stack to iterate on our experiments more quickly). I am working on refactoring this stack a bit so that we don't have this stream callback at all.

Regarding benchmarks, we've actually thoroughly benchmarked the final solution on TSC/CU workloads, and I will share a doc summarizing these findings. In summary, this entire stack introduces less than 1% QPS regression on TSC (with NE parity) and almost no QPS regression on CU.